designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,527 @@
+
+export default [
+
+    {
+        plugins: {
+            js
+        },
+        files: ["**/src/*.js"],
+        extends: ["js/recommended"]
+    }
+
+];
+```
+
+Here, `js/recommended` refers to the plugin defined as `js`. Internally, ESLint will look up the plugin with the name `js`, looks at the `configs` key, and retrieve the `recommended` key as a replacement for the string `"js/recommended"`. 
+
+This has several advantages:
+
+1. It gives ESLint knowledge of where the config is coming from. When passing in an object or an array, ESLint doesn't know the config's origin. With more information, we can give better error messages when things go wrong.
+1. The config is guaranteed to have a name (the string in `extends`) that can be used for debugging purposes.
+1. It encourages plugin authors to use the `configs` key on their plugin in order to allow this usage.
+1. It allows ESLint to modify configs before they are used (see below).
+
+#### Reassignable Plugin Configs
+
+When using named configs, this gives ESLint the opportunity to inspect and change the config before applying it. This means we have the opportunity to improve the plugin config experience.
+
+Right now, bundling a config in a plugin is a bit of a messy process for two reasons:
+
+1. You need to include the plugin in the config directly, which means that you can't just have an single object literal defined as the plugin. You need to define the plugin first and then add the configs with the plugin reference second.
+1. You don't actually know what namespace a user will assign the plugin to, so while you might assume someone will import `eslint-plugin-import` as `import` and you then refer to rules from the plugin as `import/rule`, there is nothing preventing the user from assigning the plugin to a different name. In that case, the user won't know how to override rule options or assign different languages from the plugin.
+
+Here's an actual example from [`@eslint/json`](https://github.com/eslint/json):
+
+```js
+const plugin = {
+	meta: {
+		name: "@eslint/json",
+		version: "0.6.0", // x-release-please-version
+	},
+	languages: {
+		json: new JSONLanguage({ mode: "json" }),
+		jsonc: new JSONLanguage({ mode: "jsonc" }),
+		json5: new JSONLanguage({ mode: "json5" }),
+	},
+	rules: {
+		"no-duplicate-keys": noDuplicateKeys,
+		"no-empty-keys": noEmptyKeys,
+	},
+
+    // can't include the plugin here because it must reference `plugin`
+	configs: {},
+};
+
+Object.assign(plugin.configs, {
+	recommended: {
+
+        // now we can reference plugin
+		plugins: { json: plugin },
+		rules: {
+			"json/no-duplicate-keys": "error",
+			"json/no-empty-keys": "error",
+		},
+	},
+});
+
+export default plugin;
+```
+
+Here, we are hardcoding the namespace `json` even though that might not be the namespace that the user assigns to this plugin. This is something we can now address with the use of `extends` because we have the ability to alter the config before inserting it.
```

Comment:
> ... that might not be the namespace that the user assigns to this plugin. This is something we can now address ...

I would prefer not to encourage people to rename plugins. Same reasoning as https://github.com/eslint/eslint/discussions/17766: it creates a surprising discrepancy between plugin _authors_, plugin _config re-exporters_, and end-_users_.

For example, with `@typescript-eslint/` vs. `ts/` as a prefix:

* https://github.com/TanStack/config/issues/115: users of the TanStack ESLint config were blocked from also using `tseslint.*` configs
* https://github.com/antfu/eslint-config/issues/527: intentional deviation to avoid the long name prefix

...so, as Brad predicted in https://github.com/eslint/eslint/discussions/17766#discussioncomment-9234385, there is some user request for this. But outside of heavily managed setups such as `antfu/eslint-config` I don't see any reason to encourage it in the wild.

Reply:
I think there's a difference between *allowing* this and *encouraging*. This functionality was always intended to be a feature of flat config and the ability to specify arbitrary namespaces for plugins was requested way back in eslintrc.

Moreover, the feedback we've received is that people expect this to work today, and it doesn't.

What we are going for here is trying to reduce the amount of surprises in the way the config system works. As an end user, you just want to be able to say `plugins: { foo }` and then have `rules: { "foo/bar": "error"}` just work. You don't care about anything else. Needing to ensure you're using the prescribed namespace for a plugin means needing to read yet another README, and that's what we're trying to avoid.